js/eventloop.go

```diff
@@ -0,0 +1,166 @@
+func newEventLoop(vu modules.VU) *eventLoop {
+	e := &eventLoop{
+		wakeupCh:                 make(chan struct{}, 1),
+		pendingPromiseRejections: make(map[*goja.Promise]struct{}),
+		vu:                       vu,
+	}
+	vu.Runtime().SetPromiseRejectionTracker(e.promiseRejectionTracker)
+	e.addSetTimeout()
+
+	return e
+}
+
+func (e *eventLoop) wakeup() {
+	select {
+	case e.wakeupCh <- struct{}{}:
+	default:
+	}
+}
+
+// registerCallback register that a callback will be invoked on the loop, preventing it from returning/finishing.
+// The returned function, upon invocation, will queue it's argument and wakeup the loop if needed.
+// If the eventLoop has since stopped, it will not be executed.
+// This function *must* be called from within running on the event loop, but its result can be called from anywhere
+func (e *eventLoop) registerCallback() func(func() error) {
+	e.lock.Lock()
+	e.registeredCallbacks++
+	e.lock.Unlock()
+
+	return func(f func() error) {
+		e.lock.Lock()
+		e.queue = append(e.queue, f)
+		e.registeredCallbacks--
+		e.lock.Unlock()
+		e.wakeup()
+	}
+}
+
+func (e *eventLoop) promiseRejectionTracker(p *goja.Promise, op goja.PromiseRejectionOperation) {
+	// No locking necessary here as the goja runtime will call this synchronously
+	// Read Notes on https://tc39.es/ecma262/#sec-host-promise-rejection-tracker
+	if op == goja.PromiseRejectionReject {
+		e.pendingPromiseRejections[p] = struct{}{}
+	} else { // goja.PromiseRejectionHandle so a promise that was previously rejected without handler now got one
+		delete(e.pendingPromiseRejections, p)
+	}
+}
+
+func (e *eventLoop) popAll() (queue []func() error, awaiting bool) {
+	e.lock.Lock()
+	queue = e.queue
+	e.queue = make([]func() error, 0, len(queue))
+	awaiting = e.registeredCallbacks != 0
+	e.lock.Unlock()
+	return
+}
+
+// start will run the event loop until it's empty and there are no unvoked registered callbacks
+// or a queued function returns an error. The provided firstCallback will be the first thing executed.
+// After start returns the event loop can be reused as long as waitOnRegistered is called
+func (e *eventLoop) start(firstCallback func() error) error {
+	e.queue = []func() error{firstCallback}
+	for {
+		queue, awaiting := e.popAll()
+
+		if len(queue) == 0 {
+			if !awaiting {
+				return nil
+			}
+			<-e.wakeupCh
+			continue
+		}
+
+		for _, f := range queue {
+			if err := f(); err != nil {
+				return err
+			}
+		}
+
+		// This will get a random unhandled rejection instead of the first one, for example.
+		// But that seems to be the case in other tools as well so it seems to not be that big of a problem.
+		for promise := range e.pendingPromiseRejections {
+			// TODO maybe throw the whole promise up and get make a better message outside of the event loop
+			value := promise.Result()
+			// try to get the stack without actually needing the runtime
+			// this might break in the future :(
+			if o, ok := promise.Result().(*goja.Object); ok {
```

Comment:
I didn't have the runtime here, at the time this code was written

Reply:
fixed in [11560dc](https://github.com/grafana/k6/pull/2228/commits/11560dcdcdf558cf8c2900076fe9d9792828094d) 

> Also, should we make this more explicit and type assert for *goja.Exception?

Because most of those wouldn't be exception, but [errors](https://github.com/dop251/goja/commit/45901a5728213170074cf4975efa87fc35435ee9) for which we don't have a type beside Object.